### PR TITLE
Supports incremental build in tsc --b --w mode

### PR DESCRIPTION
**Pull request by sheetalkamat**
Some notes: 
- Since we use builder, the first time program is created, its going to emit all the files and not just changed - just like it does in tsc --watch mode
- We do not cache source files yet and that's something to be looked at later.
  Two things can be done: 
   - To keep source files for program thats being updated in last few cycles
   - To release the source files if size exceeds or program is not used for few build cycles
   - Adding this to #28640
- We use hash from source file's text as the version.
- Because in tsbuild we get semantic diagnostics before emit, the `EmitAndSemanticDiagnosticsBuilderProgram`'s existing assumption was not satisfied (that is changed files are already emitted before getting semantic diagnostics). So in such case, we store list of files to emit later so as to help with that case.
- In this we also have a new Builder method to release the program and its source files to ensure that we aren't holding on program after emit and/or semantic diagnostics are complete. That way the builder state can be in memory for comparison and finding out what are new changes that need emitting while creating new program but we aren't holding onto these costly (memory wise) resources. (all the program methods are thus available only till program is released.

Here are the perf numbers:
Local change::
```
File change detected. Starting incremental compilation...

Project 'SubProject 4' is out of date because oldest output 'someOutputFile.js' is older than newest input 'SubProject 2'

Building project 'SubProject 4'...

                                Updating unchanged output timestamps of project 'SubProject 4'...

Before                          After
Files:                  8305    Files:                  8305
Lines:                509633    Lines:                509633
Nodes:               1741390    Nodes:               1741390
Identifiers:          563590    Identifiers:          563590
Symbols:              597260    Symbols:              316035
Types:                175190    Types:                   232
Memory used:        1418992K    Memory used:        1859282K
I/O Read time:         2.83s    I/O Read time:         3.85s
Parse time:            2.43s    Parse time:            3.42s
Program time:         23.45s    Program time:         33.07s
Bind time:             1.56s    Bind time:             1.71s
Check time:           11.86s    Check time:            0.04s
transformTime time:   13.72s    transformTime time:    0.25s
Source Map time:       2.58s    Source Map time:       0.01s
commentTime time:      1.46s    commentTime time:      0.00s
printTime time:       27.08s    printTime time:        0.28s
Emit time:            27.08s    Emit time:             0.28s
I/O Write time:       16.16s    I/O Write time:        0.04s
Total time:           63.96s    Total time:           35.09s
Building project 'SubProject 5'...

                                Updating unchanged output timestamps of project 'SubProject 5'...

Before                          After
Files:                  1875    Files:            1875
Lines:                207731    Lines:          207731
Nodes:                632955    Nodes:          632955
Identifiers:          212573    Identifiers:    212573
Symbols:              162334    Symbols:        125245
Types:                 22975    Types:              71
Memory used:        1095817K    Memory used:   311574K
I/O Read time:         1.13s    I/O Read time:   0.74s
Parse time:            2.03s    Parse time:      0.87s
Program time:          9.64s    Program time:    5.97s
Bind time:             0.67s    Bind time:       0.48s
Check time:            2.17s
transformTime time:    1.22s
Source Map time:       0.77s
commentTime time:      0.34s
printTime time:        4.16s
Emit time:             4.16s
I/O Write time:        1.92s
Total time:           16.64s    Total time:      6.45s
Building project 'SubProject 6'...

                                Updating unchanged output timestamps of project 'SubProject 6'...

Before                          After
Files:                  2396    Files:            2396
Lines:                221677    Lines:          221677
Nodes:                669898    Nodes:          669898
Identifiers:          221017    Identifiers:    221017
Symbols:              189874    Symbols:        143939
Types:                 30412    Types:              71
Memory used:        1442324K    Memory used:   333909K
I/O Read time:         0.92s    I/O Read time:   0.91s
Parse time:            1.40s    Parse time:      1.17s
Program time:          8.91s    Program time:    7.73s
Bind time:             0.43s    Bind time:       0.39s
Check time:            1.74s
transformTime time:    0.67s
Source Map time:       0.18s
commentTime time:      0.10s
printTime time:        1.67s
Emit time:             1.67s
I/O Write time:        1.40s
Total time:           12.74s    Total time:      8.12s
Building project 'SubProject 7'...

                                Updating unchanged output timestamps of project 'SubProject 7'...

Before                          After
Files:                  9326    Files:            9326
Lines:                552828    Lines:          552830
Nodes:               1896877    Nodes:         1896879
Identifiers:          605603    Identifiers:    605605
Symbols:              608250    Symbols:        330625
Types:                173986    Types:              71
Memory used:        2367385K    Memory used:   953536K
I/O Read time:         4.22s    I/O Read time:   3.84s
Parse time:            3.18s    Parse time:      2.02s
Program time:         36.41s    Program time:   27.79s
Bind time:             2.73s    Bind time:       1.19s
Check time:           13.15s
transformTime time:   15.36s
Source Map time:       2.35s
commentTime time:      1.52s
printTime time:       28.95s
Emit time:            28.95s
I/O Write time:       13.41s
Total time:           81.23s    Total time:     28.98s
Building project 'SubProject 8'...

                                Updating unchanged output timestamps of project 'SubProject 8'...

Before                          After
Files:                  6146    Files:             6146
Lines:                367180    Lines:           367182
Nodes:               1201703    Nodes:          1201705
Identifiers:          386452    Identifiers:     386454
Symbols:              406182    Symbols:         232990
Types:                 94077    Types:               71
Memory used:        1704251K    Memory used:   1333457K
I/O Read time:         3.06s    I/O Read time:    2.50s
Parse time:            3.61s    Parse time:       2.06s
Program time:         31.86s    Program time:    19.05s
Bind time:             1.76s    Bind time:        0.77s
Check time:           10.90s
transformTime time:    7.64s
Source Map time:       0.96s
commentTime time:      0.91s
printTime time:       13.81s
Emit time:            13.81s
I/O Write time:        6.67s
Total time:           58.33s    Total time:      19.82s
Building project 'SubProject 9'...

                                Updating unchanged output timestamps of project 'SubProject 9'...

Before                          After
Files:                  3781    Files:             3781
Lines:                264899    Lines:           264899
Nodes:                834360    Nodes:           834360
Identifiers:          278052    Identifiers:     278052
Symbols:              261052    Symbols:         178924
Types:                 52797    Types:               71
Memory used:        2175398K    Memory used:   1620896K
I/O Read time:         1.67s    I/O Read time:    1.51s
Parse time:            1.98s    Parse time:       1.19s
Program time:         16.03s    Program time:    11.75s
Bind time:             0.77s    Bind time:        0.49s
Check time:            4.86s
transformTime time:    4.24s
Source Map time:       0.40s
commentTime time:      0.21s
printTime time:        6.63s
Emit time:             6.63s
I/O Write time:        2.03s
Total time:           28.29s    Total time:      12.24s


Solution Total Time:
Before: 261.19, After: 110.7
```

Local change:
```
File change detected. Starting incremental compilation...

Project 'SubProject 4' is out of date because oldest output 'someOutputFile.js' is older than newest input 'SubProject 2'

Building project 'SubProject 4'...

                                Updating unchanged output timestamps of project 'SubProject 4'...

Before                          After
Files:                  8305    Files:                 8305
Lines:                509635    Lines:               509635
Nodes:               1741406    Nodes:              1741406
Identifiers:          563597    Identifiers:         563597
Symbols:              597262    Symbols:             315957
Types:                175191    Types:                  114
Memory used:        3243603K    Memory used:        690192K
I/O Read time:         3.23s    I/O Read time:        3.28s
Parse time:            4.00s    Parse time:           2.31s
Program time:         35.91s    Program time:        24.59s
Bind time:             1.75s    Bind time:            1.27s
Check time:           15.76s    Check time:           0.01s
transformTime time:   15.10s    transformTime time:   0.00s
Source Map time:       2.93s    Source Map time:      0.00s
commentTime time:      1.85s    commentTime time:     0.00s
printTime time:       30.68s    printTime time:       0.00s
Emit time:            30.68s    Emit time:            0.00s
I/O Write time:       14.47s    I/O Write time:       0.00s
Total time:           84.10s    Total time:          25.87s
                                Project 'SubProject 5' is up to date with .d.ts files from its dependencies

                                Updating output timestamps of project 'SubProject 5'...

                                Project 'SubProject 6' is up to date with .d.ts files from its dependencies

                                Updating output timestamps of project 'SubProject 6'...

                                Project 'SubProject 7' is up to date with .d.ts files from its dependencies

                                Updating output timestamps of project 'SubProject 7'...

                                Project 'SubProject 8' is up to date with .d.ts files from its dependencies

                                Updating output timestamps of project 'SubProject 8'...

                                Project 'SubProject 9' is up to date with .d.ts files from its dependencies

                                Updating output timestamps of project 'SubProject 9'...


Solution Total Time:
Before: 84.1, After: 25.87
Found 0 errors. Watching for file changes.
```

```

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -3101,18 +3108,28 @@ namespace ts {
...
     /* @internal */
-    export function parseConfigHostFromCompilerHost(host: CompilerHost): ParseConfigFileHost {
+    export function parseConfigHostFromCompilerHostLike(host: CompilerHostLike, directoryStructureHost: DirectoryStructureHost = host): ParseConfigFileHost {
         return {
             fileExists: f => host.fileExists(f),
             readDirectory(root, extensions, excludes, includes, depth) {
-                Debug.assertDefined(host.readDirectory, "'CompilerHost.readDirectory' must be implemented to correctly process 'projectReferences'");
-                return host.readDirectory!(root, extensions, excludes, includes, depth);
+                Debug.assertDefined(directoryStructureHost.readDirectory, "'CompilerHost.readDirectory' must be implemented to correctly process 'projectReferences'");
+                return directoryStructureHost.readDirectory!(root, extensions, excludes, includes, depth);
```
Comment (ajafff) [inline src/compiler/program.ts]: why is `directoryStructureHost` only used for `readDirectory`? Shouldn't it also be used for `readFile` and `fileExists`?

**Comment by sheetalkamat** [comment]
Ping @sandersn @rbuckton @weswigham @DanielRosenwasser Would like to get this in soon.

File: src/compiler/core.ts [hunk 2]
```diff
@@ -1387,6 +1391,14 @@ namespace ts {
         return result;
     }
 
+    export function copyProperities<T1 extends T2, T2>(first: T1, second: T2) {
```
Comment (sandersn) [inline src/compiler/core.ts]: typo:Properties

File: src/compiler/sys.ts [hunk 3]
```diff
@@ -2,6 +2,16 @@ declare function setTimeout(handler: (...args: any[]) => void, timeout: number):
 declare function clearTimeout(handle: any): void;
 
 namespace ts {
+    /**
+     * djb2 hashing algorithm
+     * http://www.cse.yorku.ca/~oz/hash.html
+     */
+    /* @internal */
+    export function generateDjb2Hash(data: string): string {
+        const chars = data.split("").map(str => str.charCodeAt(0));
```
Comment (sandersn) [inline src/compiler/sys.ts]: Is this used a lot? It seems like a for loop over the character indices would allocate less than data.split.map. Of course the optimizer might do the right thing.

Comment (sandersn) [reply src/compiler/sys.ts]: Later: oh, this is just a code move.

Comment (sheetalkamat) [reply src/compiler/sys.ts]: Note that this is a fallback for creating hash to store instead of storing text of generated .d.ts files signature. With this change it would also be used as backup (when node api doesn't support creating hash) to get the hash for source File text as sourceFile's version

File: src/compiler/builder.ts [hunk 4]
```diff
@@ -132,6 +157,38 @@ namespace ts {
...
+    function releaseCache(state: BuilderProgramState) {
+        BuilderState.releaseCache(state);
+        state.program = undefined;
+    }
+
+    /**
+     * Creates a clone of the state
+     */
+    function cloneBuilderProgramState(state: Readonly<BuilderProgramState>): BuilderProgramState {
+        const newState = BuilderState.clone(state) as BuilderProgramState;
```
Comment (sandersn) [inline src/compiler/builder.ts]: This seems weird -- if BuilderState can clone itself, why do we have to copy a bunch of properties here too?

File: src/compiler/builder.ts [hunk 5]
```diff
@@ -182,24 +239,46 @@ namespace ts {
...
     }
 
+    /**
+     * Returns next file to be emitted from files that retrieved semantic diagnostics but did not emit yet
+     */
+    function getNextAffectedFilePendingEmit(state: BuilderProgramState): SourceFile | undefined {
+        const { affectedFilesPendingEmit } = state;
+        if (affectedFilesPendingEmit) {
+            const seenEmittedFiles = state.seenEmittedFiles || (state.seenEmittedFiles = createMap());
+            for (let affectedFilesIndex = state.affectedFilesPendingEmitIndex!; affectedFilesIndex < affectedFilesPendingEmit.length; affectedFilesIndex++) {
```
Comment (sandersn) [inline src/compiler/builder.ts]: I think `i` would be a more readable name here.

File: src/compiler/builder.ts (hunk 4 above)
Comment (sheetalkamat) [reply src/compiler/builder.ts]: Because you cloned BuilderState and this is BuilderProgramState (interface BuilderProgramState extends BuilderState)

Comment (sheetalkamat) [reply src/compiler/builder.ts]: I mean these are explicit properties in `BuilderProgramState`

File: src/compiler/builder.ts [hunk 6]
```diff
@@ -560,33 +670,66 @@ namespace ts {
...
+            getState: notImplemented,
+            backupCurrentState: noop,
+            useBackupState: noop,
+            getProgram: () => Debug.assertDefined(state.program),
+            getProgramOrUndefined: () => state.program,
+            releaseProgram: () => state.program = undefined,
+            getCompilerOptions: () => state.compilerOptions,
+            getSourceFile: fileName => Debug.assertDefined(state.program).getSourceFile(fileName),
+            getSourceFiles: () => Debug.assertDefined(state.program).getSourceFiles(),
+            getOptionsDiagnostics: cancellationToken => Debug.assertDefined(state.program).getOptionsDiagnostics(cancellationToken),
```
Comment (sandersn) [inline src/compiler/builder.ts]: will `program` change between defined and undefined? Is it possible to just capture `const program = Debug.assertDefined(state.program)` as the first line of `createRedirectedBuilderProgram`?

Comment (sheetalkamat) [reply src/compiler/builder.ts]: Hmm I thought I had mentioned this in descriptions but looks like I missed it.  Sorry about that. I will update it there too.
The main intention of not capturing program is that we want to release program and its source files once the errors and emit is complete so that in ts build scenario we aren't holding onto bunch of programs and their sourcefiles for forever. And hence the asserts that these are called when program is available. The methods to create new BuildProgram from this old state shouldn't depend on having program around.

File: src/compiler/builder.ts [hunk 7]
```diff
@@ -461,18 +550,39 @@ namespace ts {
          * in that order would be used to write the files
          */
         function emitNextAffectedFile(writeFile?: WriteFileCallback, cancellationToken?: CancellationToken, emitOnlyDtsFiles?: boolean, customTransformers?: CustomTransformers): AffectedFileResult<EmitResult> {
-            const affected = getNextAffectedFile(state, cancellationToken, computeHash);
+            let affected = getNextAffectedFile(state, cancellationToken, computeHash);
+            let isPendingEmitFile = false;
             if (!affected) {
```
Comment (sandersn) [inline src/compiler/builder.ts]: Could this whole block be moved into `getNextAffectedFile`? The logic seems a little out of place here. The main obstacle I see is that the return type would need to include `isPendingEmitFile`, or it would have to be calculated from `affected`.

File: src/compiler/builder.ts [hunk 8]
```diff
@@ -560,33 +670,66 @@ namespace ts {
...
+                if (affected !== state.program && kind === BuilderProgramKind.EmitAndSemanticDiagnosticsBuilderProgram) {
+                    (affectedFilesPendingEmit || (affectedFilesPendingEmit = [])).push((affected as SourceFile).path);
+                }
+                doneWithAffectedFile(state, affected);
+            }
+
+            // In case of emit builder, cache the files to be emitted
+            if (affectedFilesPendingEmit) {
+                state.affectedFilesPendingEmit = concatenate(state.affectedFilesPendingEmit, affectedFilesPendingEmit);
+                if (state.affectedFilesPendingEmitIndex === undefined) {
```
Comment (sandersn) [inline src/compiler/builder.ts]: can this ever be true? affectedFilePendingEmit is just checked to be defined, so at least it should be defined with length 0. Also, what is affectedFilePendingEmitIndex set to in the false case?

Comment (sheetalkamat) [reply src/compiler/builder.ts]: Its not currently true (since we are synchronized that is all semantic diagnostics are done and then emit is done. but with future plan to create API out of it where the operations are granular) I have this check here..  
`affectedFilesPendingEmitIndex  === undefined` if `state.affectedFilesPendingEmit` before addending current set of affected Files is undefined. (in short all the files are iterated). 
In false branch we retain the index because we want to emit from the existing index to Last added files (note how these are concatenated and used during emit)

File: src/compiler/program.ts [hunk 9]
```diff
@@ -203,18 +203,25 @@ namespace ts {
         return compilerHost;
     }
 
+    interface ComplierHostLikeForCache {
```
Comment (sandersn) [inline src/compiler/program.ts]: typo:CompilerHostLikeForCache

Comment (sandersn) [reply src/compiler/program.ts]: Also it would be nice to put all the CompilerHost-related types in one place and try to come up with meaningful names for them, or least relationships.

File: src/compiler/sys.ts (hunk 3 above)
Comment (sandersn) [reply src/compiler/sys.ts]: Seems expensive for what it does, but not worth changing right now I think.

File: src/compiler/program.ts (hunk 9 above)
Comment (sheetalkamat) [reply src/compiler/program.ts]: good suggestion. Will add that to #28640

File: src/compiler/sys.ts (hunk 3 above)
Comment (sheetalkamat) [reply src/compiler/sys.ts]: Yeah thats the default function that is independent of this PR.

File: src/compiler/tsbuild.ts [hunk 10]
```diff
@@ -492,6 +497,27 @@ namespace ts {
             clearMap(allWatchedWildcardDirectories, wildCardWatches => clearMap(wildCardWatches, closeFileWatcherOf));
             clearMap(allWatchedInputFiles, inputFileWatches => clearMap(inputFileWatches, closeFileWatcher));
             clearMap(allWatchedConfigFiles, closeFileWatcher);
+            builderPrograms.clear();
+            updateGetSourceFile();
```
Comment (sandersn) [inline src/compiler/tsbuild.ts]: why do we have to do this?

Comment (sandersn) [reply src/compiler/tsbuild.ts]: It seems like it might be related to your comments on EmitAndSemanticDiagnosticsBuilderProgram, but I don't understand how.

File: src/compiler/tsbuild.ts [hunk 11]
```diff
@@ -944,16 +990,40 @@ namespace ts {
...
-            for (const project of dependencyGraph.buildQueue) {
+            for (let index = buildQueue.indexOf(resolved) + 1; index < buildQueue.length; index++) {
+                const project = buildQueue[index];
                 const prepend = referencingProjects.getValue(project);
-                // If the project is referenced with prepend, always build downstream projectm,
-                // otherwise queue it only if declaration output changed
-                if (prepend || (prepend !== undefined && !(buildResult & BuildResultFlags.DeclarationOutputUnchanged))) {
+                if (prepend !== undefined) {
+                    // If the project is referenced with prepend, always build downstream project,
+                    // If declaration output is changed changed, build the project
```
Comment (sandersn) [inline src/compiler/tsbuild.ts]: typo:changed changed

File: src/compiler/tsbuild.ts [hunk 12]
```diff
@@ -944,16 +990,40 @@ namespace ts {
                 return;
             }
 
+            if (status.type === UpToDateStatusType.UpToDateWithUpstreamTypes) {
```
Comment (sandersn) [inline src/compiler/tsbuild.ts]: how is this new status related to watch mode? It sounds unrelated.

File: src/compiler/tsbuild.ts (hunk 10 above)
Comment (sheetalkamat) [reply src/compiler/tsbuild.ts]: Currently `resetBuildContext` is API exposed. We are yet to see if that's really needed or is it test only but that api just means clear and be in initial state as if just created.

File: src/compiler/tsbuild.ts [hunk 13]
```diff
@@ -944,16 +990,40 @@ namespace ts {
...
+                updateOutputTimestamps(proj);
+                return;
+            }
+
             const buildResult = buildSingleProject(resolved);
-            const dependencyGraph = getGlobalDependencyGraph();
-            const referencingProjects = dependencyGraph.referencingProjectsMap.getValue(resolved);
+            if (buildResult & BuildResultFlags.AnyErrors) return;
+
+            const { referencingProjectsMap, buildQueue } = getGlobalDependencyGraph();
```
Comment (sandersn) [inline src/compiler/tsbuild.ts]: is the first item returned not a dependencyGraph anymore? It seems to be used the same as before.

File: src/compiler/tsbuild.ts [hunk 14]
```diff
@@ -944,16 +990,40 @@ namespace ts {
...
             // Always use build order to queue projects
-            for (const project of dependencyGraph.buildQueue) {
+            for (let index = buildQueue.indexOf(resolved) + 1; index < buildQueue.length; index++) {
+                const project = buildQueue[index];
                 const prepend = referencingProjects.getValue(project);
-                // If the project is referenced with prepend, always build downstream projectm,
-                // otherwise queue it only if declaration output changed
-                if (prepend || (prepend !== undefined && !(buildResult & BuildResultFlags.DeclarationOutputUnchanged))) {
+                if (prepend !== undefined) {
+                    // If the project is referenced with prepend, always build downstream project,
```
Comment (sandersn) [inline src/compiler/tsbuild.ts]: the last 'project' should be plural 'projects'

File: src/compiler/tsbuild.ts (hunk 12 above)
Comment (sheetalkamat) [reply src/compiler/tsbuild.ts]: If you have `proj` depending on `dep` when change happens to the `dep` thats local (not changing its d.ts file) the file will get updated stamp and hence the status of this build would be `UpToDateWithUpstreamTypes` because of our own knowledge that the change was local and .d.ts not really changed. But if we did not update the type stamps, if we kill build at this point and new process starts, it wont have that info and it would create new program and emit etc. So this was existing but that got fixed as part of making sure dependent projects have correct time stamps in case of incremental build (Note how we have to update timestamps of unchanged emit files when doing incremental emit to ensure it doesnt falsely assumed to be not uptodate when new process with tsc -b starts where this info about the d.ts updates etc is not in the memory)

Comment (sheetalkamat) [reply src/compiler/tsbuild.ts]: So this is more about making sure that incrementally built and done projects dont unnecessarily start creating and emitting when tsc -b is invoked in another process

File: src/compiler/tsbuild.ts (hunk 13 above)
Comment (sheetalkamat) [reply src/compiler/tsbuild.ts]: Its always been these two things returned, Now we are just replacing it with destructuring of these since we use both of these anyways.

File: src/compiler/tsbuild.ts [hunk 15]
```diff
@@ -1311,6 +1411,20 @@ namespace ts {
...
+            return first(getOutFileOutputs(project));
+        }
+
+        for (const inputFile of project.fileNames) {
+            const outputs = getOutputFileNames(inputFile, project);
+            if (outputs.length) {
+                return first(outputs);
+            }
+        }
+        return Debug.fail(`project ${project.options.configFilePath} expected to have atleast one output`);
```
Comment (sandersn) [inline src/compiler/tsbuild.ts]: typo: at least

File: src/compiler/watch.ts [hunk 16]
```diff
@@ -194,23 +194,105 @@ namespace ts {
...
+    export const enum WatchType {
+        ConfigFile = "Config file",
+        SourceFile = "Source file",
+        MissingFile = "Missing file",
+        WildcardDirectory = "Wild card directory",
+        FailedLookupLocations = "Failed Lookup Locations",
+        TypeRoots = "Type roots"
+    }
+
+    interface WatchFactory<X, Y= undefined> extends ts.WatchFactory<X, Y> {
```
Comment (sandersn) [inline src/compiler/watch.ts]: should probably have a space between `Y=`.

File: src/compiler/watch.ts [hunk 17]
```diff
@@ -194,23 +194,105 @@ namespace ts {
...
+        watchLogLevel: WatchLogLevel;
+        writeLog: (s: string) => void;
+    }
+
+    export function createWatchFactory<Y = undefined>(host: { trace?(s: string): void; }, options: { extendedDiagnostics?: boolean; diagnostics?: boolean; }) {
+        const watchLogLevel = host.trace ? options.extendedDiagnostics ? WatchLogLevel.Verbose :
```
Comment (sandersn) [inline src/compiler/watch.ts]: this formatting is misleading since the inner-wrapped flat series of conditionals has a line-break in the middle. Is there a way to make the whole conditional a flat series of checks? That would be easier to read than the nested `host.trace ? options.extendedDiagnostics ? ...` style.

File: src/compiler/watch.ts [hunk 18]
```diff
@@ -194,23 +194,105 @@ namespace ts {
...
+            options.diagnostics ? WatchLogLevel.TriggerOnly : WatchLogLevel.None : WatchLogLevel.None;
+        const writeLog: (s: string) => void = watchLogLevel !== WatchLogLevel.None ? (s => host.trace!(s)) : noop;
+        const result = getWatchFactory<WatchType, Y>(watchLogLevel, writeLog) as WatchFactory<WatchType, Y>;
```
Comment (sandersn) [inline src/compiler/watch.ts]: It's weird that we pass watchLogLevel and writeLog to getWatchFactory and still have to set them on the result right after.

File: src/compiler/watch.ts [hunk 19]
```diff
@@ -194,23 +194,102 @@ namespace ts {
...
+            getDefaultLibLocation: host.getDefaultLibLocation && (() => host.getDefaultLibLocation!()),
+            getDefaultLibFileName: options => host.getDefaultLibFileName(options),
+            writeFile,
+            getCurrentDirectory: memoize(() => host.getCurrentDirectory()),
+            useCaseSensitiveFileNames: () => useCaseSensitiveFileNames,
+            getCanonicalFileName: createGetCanonicalFileName(useCaseSensitiveFileNames),
+            getNewLine: memoize(() => getNewLineCharacter(getCompilerOptions(), () => host.getNewLine())),
+            fileExists: f => host.fileExists(f),
+            readFile: f => host.readFile(f),
+            trace: host.trace && (s => host.trace!(s)),
```
Comment (rbuckton) [inline src/compiler/watch.ts]: I've seen a fair number of similar expressions like this. Would it make sense to generalize this? Something along the lines of:

```ts
function maybeBind<T, A extends any[], R>(obj: T, fn: ((this: T, ...args: A) => R) | undefined): ((...args: A) => R) | undefined {
  return fn ? fn.bind(obj) : undefined;
}
```

File: src/compiler/watch.ts [hunk 20]
```diff
@@ -708,7 +758,7 @@ namespace ts {
 
             // Create new source file if requested or the versions dont match
             if (!hostSourceFile || shouldCreateNewSourceFile || !isFilePresentOnHost(hostSourceFile) || hostSourceFile.version.toString() !== hostSourceFile.sourceFile.version) {
-                const sourceFile = getNewSourceFile();
+                const sourceFile = getNewSourceFile.call(compilerHost, fileName, languageVersion, onError);
```
Comment (rbuckton) [inline src/compiler/watch.ts]: The host we create from `createCompilerHostFromProgramHost` doesn't use `this` (all of its methods are arrows). Do we need the indirect call via `Function#call` here?

File: src/compiler/program.ts [hunk 21]
```diff
@@ -2187,7 +2187,7 @@ namespace ts {
         }
 
         function createRedirectSourceFile(redirectTarget: SourceFile, unredirected: SourceFile, fileName: string, path: Path, resolvedPath: Path, originalFileName: string): SourceFile {
-            const redirect = createRedirectObject(redirectTarget);
+            const redirect = Object.create(redirectTarget);
```
Comment (rbuckton) [inline src/compiler/program.ts]: `Object.create` returns `any`, so we should add a type annotation to `redirect` for type safety.

File: src/compiler/builder.ts [hunk 22]
```diff
@@ -614,10 +765,24 @@ namespace ts {
     export interface BuilderProgram {
         /*@internal*/
         getState(): BuilderProgramState;
+        /*@internal*/
+        backupCurrentState(): void;
```
Comment (rbuckton) [inline src/compiler/builder.ts]: minor nit (up to you if you want to change this): Would `backupState`/`restoreState` be clearer?

**Comment by sheetalkamat** [comment]
@sandersn comments?

**Comment by xaviergonz** [comment]
Just wondering, wouldn't it be even faster if the cache was also based on file size as first check before doing the actual hash?
